src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_SwitchExpression.cs

```diff
@@ -16,43 +16,52 @@ internal partial class LocalRewriter
         {
-            // The switch expression is lowered to an expression that involves the use of side-effects
-            // such as jumps and labels, therefore it is represented by a BoundSpillSequence and
-            // the resulting nodes will need to be "spilled" to move such statements to the top
-            // level (i.e. into the enclosing statement list).
-            this._needsSpilling = true;
             return SwitchExpressionLocalRewriter.Rewrite(this, node);
         }
 
         private sealed class SwitchExpressionLocalRewriter : BaseSwitchLocalRewriter
         {
-            private SwitchExpressionLocalRewriter(BoundConvertedSwitchExpression node, LocalRewriter localRewriter)
+            private SwitchExpressionLocalRewriter(BoundConvertedSwitchExpression node, LocalRewriter localRewriter, bool generateInstrumentation)
                 : base(node.Syntax, localRewriter, node.SwitchArms.SelectAsArray(arm => arm.Syntax),
-                      generateInstrumentation: !node.WasCompilerGenerated && localRewriter.Instrument)
+                      generateInstrumentation: generateInstrumentation)
             {
             }
 
             public static BoundExpression Rewrite(LocalRewriter localRewriter, BoundConvertedSwitchExpression node)
             {
-                var rewriter = new SwitchExpressionLocalRewriter(node, localRewriter);
-                BoundExpression result = rewriter.LowerSwitchExpression(node);
-                rewriter.Free();
-                return result;
+                var generateInstrumentation = !node.WasCompilerGenerated && localRewriter.Instrument;
+                // When compiling for Debug (not Release), we produce the most detailed sequence points.
+                var produceDetailedSequencePoints = generateInstrumentation && localRewriter._compilation.Options.OptimizationLevel != OptimizationLevel.Release;
+
+                if (produceDetailedSequencePoints)
+                {
+                    localRewriter._needsSpilling = true;
+                    var rewriter = new SwitchExpressionLocalRewriter(node, localRewriter, generateInstrumentation: true);
+                    BoundExpression result = rewriter.LowerSwitchExpressionForDebug(node);
+                    rewriter.Free();
+                    return result;
+                }
+                else
+                {
+                    var rewriter = new SwitchExpressionLocalRewriter(node, localRewriter, generateInstrumentation: false);
+                    BoundExpression result = rewriter.LowerSwitchExpressionForRelease(node);
+                    rewriter.Free();
+                    return result;
+                }
             }
 
-            private BoundExpression LowerSwitchExpression(BoundConvertedSwitchExpression node)
+            private BoundExpression LowerSwitchExpressionForDebug(BoundConvertedSwitchExpression node)
```

Comment:
From [Expression Breakpoints.md](https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Expression%20Breakpoints.md)
> Sequence points are constrained to appear only at locations in the code where the evaluation stack is empty

I've separated the lowering for debug/release, not sure if it's possible to combine the two somehow.